### PR TITLE
Adding netmhcpan-3.0.

### LINKED ISSUE #73: Add netMHCpan 3.0

### PR DESCRIPTION
**Pull request by julia326**
Fixes #73, #71 

This change is [Reviewable](https://reviewable.io/reviews/hammerlab/mhctools/72)

### CODE REVIEW

File: mhctools/__init__.py [hunk 1]
```diff
@@ -32,6 +34,8 @@
     "NetMHC4",
     "NetMHCcons",
     "NetMHCpan",
+    "NetMHCpan2",
```
Comment (tavinathanson) [inline mhctools/__init__.py]: Maybe call this 2.8 to avoid confusing with e.g. older versions like 2.4?

File: mhctools/file_formats.py [hunk 2]
```diff
@@ -235,7 +235,7 @@ def parse_netmhc4_stdout(
         rank_index=13,
         log_ic50_index=11)
 
-def parse_netmhcpan_stdout(
+def parse_netmhcpan2_stdout(
         stdout,
         fasta_dictionary,
         prediction_method_name="netmhcpan",
```
Comment (tavinathanson) [inline mhctools/file_formats.py:253]: Call this `netmhcpan28`?

File: mhctools/file_formats.py [hunk 3]
```diff
@@ -274,6 +274,33 @@ def parse_netmhcpan_stdout(
...
+    1  HLA-B*18:01        MFCQLAKT  MFCQLAKT-  0  0  0  8  1     MFCQLAKT     sequence0_0 0.02864 36676.0   45.00
+    2  HLA-B*18:01        FCQLAKTY  F-CQLAKTY  0  0  0  1  1     FCQLAKTY     sequence0_0 0.07993 21056.5   13.00
+    """
+    return parse_stdout(
+        stdout=stdout,
+        fasta_dictionary=fasta_dictionary,
+        prediction_method_name=prediction_method_name,
+        sequence_key_mapping=sequence_key_mapping,
+        key_index=10,
+        offset_index=4,
```
Comment (tavinathanson) [inline mhctools/file_formats.py]: We talked about `offset` in `mhctools` being equal to `pos`, right? AFAIK this should break things. Does it? (If not, we should have tests to catch this?)

File: mhctools/netmhc_cons.py [hunk 4]
```diff
@@ -30,7 +30,7 @@ def __init__(
             program_name=program_name,
             alleles=alleles,
             epitope_lengths=epitope_lengths,
-            parse_output_fn=parse_netmhcpan_stdout,
+            parse_output_fn=parse_netmhccons_stdout,
```
Comment (tavinathanson) [inline mhctools/netmhc_cons.py:33]: Thanks for fixing!

File: mhctools/netmhc_pan.py [hunk 5]
```diff
@@ -13,27 +13,42 @@
...
+
+logger = logging.getLogger(__name__)
+
+
+def NetMHCpan(alleles,
+           epitope_lengths=[9],
+           program_name="netMHCpan",
+           max_file_records=None):
+    """
+    This function wraps NetMHCPan2 and NetMHCPan3 to automatically detect which class
```
Comment (tavinathanson) [inline mhctools/netmhc_pan.py]: Nit: `NetMHCpan2` and `NetMHCpan3` vs. `NetMHCPan2` and `NetMHCPan3`

File: mhctools/netmhc_pan.py [hunk 6]
```diff
@@ -13,27 +13,42 @@
...
+    to use, with the help of the miraculous and strange '--version' netmhcpan argument.
+    """
+    # convert to str since Python3 returns a `bytes` object
+    output = check_output([
+        program_name, "--version", "3",
```
Comment (tavinathanson) [inline mhctools/netmhc_pan.py:39]: Clarify here that "3" is meaningless?

File: mhctools/netmhc_pan.py [hunk 7]
```diff
@@ -13,27 +13,42 @@
...
+    output_str = output.decode("ascii", "ignore")
+    if "NetMHCpan version 2.8" in output_str:
+    	return NetMHCpan2(
+	    	alleles=alleles,
+	    	epitope_lengths=epitope_lengths,
+	    	program_name=program_name,
+	    	max_file_records=max_file_records)
+
+    if "NetMHCpan version 3.0" not in output_str:
+    	logger.info("Unknown version of netmhcpan, defaulting to 3.0")
```
Comment (tavinathanson) [inline mhctools/netmhc_pan.py]: Probably raise something here vs. continuing?

Comment (tavinathanson) [inline mhctools/netmhc_pan.py]: Nit: `NetMHCpan` is the correct capitalization of the tool

File: mhctools/netmhc_pan2.py [hunk 8]
```diff
@@ -0,0 +1,41 @@
...
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+
+from __future__ import print_function, division, absolute_import
+
+from .base_commandline_predictor import BaseCommandlinePredictor
+from .file_formats import parse_netmhcpan2_stdout
+
+class NetMHCpan2(BaseCommandlinePredictor):
```
Comment (tavinathanson) [inline mhctools/netmhc_pan2.py]: Same here, maybe call `NetMHCpan28`?

File: test/test_netmhc_pan.py [hunk 9]
```diff
@@ -1,20 +0,0 @@
-from mhctools import NetMHCpan
-from mhctools.alleles import normalize_allele_name
-
-
-DEFAULT_ALLELE = 'HLA-A*02:01'
-
-def test_netmhc_pan():
```
Comment (tavinathanson) [inline test/test_netmhc_pan.py:7]: Where'd this go again?

**Review by tavinathanson — CHANGES_REQUESTED** [review]
Looks good, just a few comments/questions!

File: mhctools/file_formats.py (hunk 3 above)
Comment (julia326) [reply mhctools/file_formats.py]: I have tests now, also specific to that field being correct in value - update coming up

File: test/test_netmhc_pan.py (hunk 9 above)
Comment (julia326) [reply test/test_netmhc_pan.py:7]: Renamed the test file, but a later change made that rename unnecessary so will bring it back in the next commit

File: test/test_netmhc_pan.py [hunk 10]
```diff
@@ -1,20 +1,34 @@
...
+        program_name=mhcpan_executable)
     fasta_dictionary = {
         "SMAD4-001": "ASIINFKELA",
         "TP53-001": "ASILLLVFYW"
     }
-    epitope_collection = pan_predictor.predict(
-        fasta_dictionary=fasta_dictionary)
+    return predictor.predict(fasta_dictionary=fasta_dictionary)
 
+def test_netmhc_pan():
```
Comment (tavinathanson) [inline test/test_netmhc_pan.py:22]: So this failed before your transform right?

**Review by tavinathanson — APPROVED** [review]
LGTM! I think it'd be great if you renamed `netmhc_pan2` to `28` as well for consistency but up to you

File: test/test_netmhc_pan.py (hunk 10 above)
Comment (julia326) [reply test/test_netmhc_pan.py:22]: It did fail, in the expected way

**Pull request merged**